Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m

```diff
@@ -0,0 +1,1083 @@
+    cellSize.height += quotedMessageSize.height;
+
+    CGSize mediaContentSize = [self bodyMediaSizeForContentWidth:contentWidth];
+    cellSize.width = MAX(cellSize.width, mediaContentSize.width);
+    cellSize.height += mediaContentSize.height;
+
+    CGSize textContentSize = [self bodyTextSizeForContentWidth:contentWidth includeMargins:YES];
+    cellSize.width = MAX(cellSize.width, textContentSize.width);
+    cellSize.height += textContentSize.height;
+
+    OWSAssert(cellSize.width > 0 && cellSize.height > 0);
+
+    if (self.hasTapForMore) {
+        cellSize.height += self.tapForMoreHeight;
+    }
+
+    cellSize = CGSizeCeil(cellSize);
+
+    return cellSize;
+}
+
+- (UIFont *)tapForMoreFont
+{
+    return [UIFont ows_regularFontWithSize:12.f];
+}
+
+- (CGFloat)tapForMoreHeight
+{
+    return (CGFloat)ceil([self tapForMoreFont].lineHeight * 1.25);
+}
+
+#pragma mark -
+
+- (CGFloat)textLeadingMargin
+{
+    CGFloat result = kBubbleTextHInset;
+    if (self.isIncoming) {
+        result += kBubbleThornSideInset;
+    }
+    return result;
+}
+
+- (CGFloat)textTrailingMargin
+{
+    CGFloat result = kBubbleTextHInset;
+    if (!self.isIncoming) {
+        result += kBubbleThornSideInset;
+    }
+    return result;
+}
+
+- (CGFloat)textTopMargin
+{
+    return kBubbleTextVInset;
+}
+
+- (CGFloat)textBottomMargin
+{
+    return kBubbleTextVInset + kBubbleThornVInset;
+}
+
+- (UIColor *)bodyTextColor
+{
+    return self.isIncoming ? [UIColor blackColor] : [UIColor whiteColor];
+}
+
+- (BOOL)isMediaBeingSent
+{
+    if (self.isIncoming) {
+        return NO;
+    }
+    if (self.cellType == OWSMessageCellType_DownloadingAttachment) {
+        return NO;
+    }
+    if (!self.attachmentStream) {
+        return NO;
+    }
+    TSOutgoingMessage *outgoingMessage = (TSOutgoingMessage *)self.viewItem.interaction;
+    return outgoingMessage.messageState == TSOutgoingMessageStateAttemptingOut;
+}
+
+- (OWSMessagesBubbleImageFactory *)bubbleFactory
+{
+    return [OWSMessagesBubbleImageFactory shared];
+}
+
+- (void)prepareForReuse
+{
+    [NSLayoutConstraint deactivateConstraints:self.viewConstraints];
+    self.viewConstraints = [NSMutableArray new];
+
+    [self.bodyTextView removeFromSuperview];
+    self.bodyTextView.text = nil;
+    self.bodyTextView.hidden = YES;
+
+    self.bubbleView.bubbleColor = nil;
+    [self.bubbleView clearPartnerViews];
+
+    for (UIView *subview in self.bubbleView.subviews) {
+        [subview removeFromSuperview];
+    }
+
+    if (self.unloadCellContentBlock) {
+        self.unloadCellContentBlock();
+    }
+    self.loadCellContentBlock = nil;
+    self.unloadCellContentBlock = nil;
+
+    [self.lastBodyMediaView removeFromSuperview];
+    self.lastBodyMediaView = nil;
+
+    [self.lastQuotedMessageView removeFromSuperview];
+    self.lastQuotedMessageView = nil;
+}
+
+#pragma mark - Gestures
+
+- (OWSMessageGestureLocation)gestureLocationForLocation:(CGPoint)locationInMessageBubble
```

Comment:
This method was added: it identifies which message content a given gesture applies to.